// @miksu/prettier/lib/language-js/index.js
"use strict";

var estreePrinter = require("./printer-estree");

var estreeJsonPrinter = require("./printer-estree-json");

var options = require("./options");

var createLanguage = require("../utils/create-language");

var languages = [createLanguage(require("linguist-languages/data/JavaScript"), {
  override: {
    since: "0.0.0",
    parsers: ["babel", "flow"],
    vscodeLanguageIds: ["javascript"]
  },
  extend: {
    interpreters: ["nodejs"]
  }
}), createLanguage(require("linguist-languages/data/JavaScript"), {
  override: {
    name: "Flow",
    since: "0.0.0",
    parsers: ["babel", "flow"],
    vscodeLanguageIds: ["javascript"],
    aliases: [],
    filenames: [],
    extensions: [".js.flow"]
  }
}), createLanguage(require("linguist-languages/data/JSX"), {
  override: {
    since: "0.0.0",
    parsers: ["babel", "flow"],
    vscodeLanguageIds: ["javascriptreact"]
  }
})];
var printers = {
  estree: estreePrinter,
  "estree-json": estreeJsonPrinter
};
module.exports = {
  languages: languages,
  options: options,
  printers: printers
};
